### PR TITLE
Testing for add, added comments, minor changes

### PR DESCRIPTION
**Pull request by krshrimali**
This PR will ideally add `ref` argument to `OpInfo` base class. The idea is to add reference checks for all the ops _eligible_. For more discussion, please check `#58294`

* [x] Migrate (but not removing yet) and modify helper functions from `UnaryUfuncOpInfo` class to `OpInfo` base class.
* [ ] Test the reference checks for multiple ops. (also decide a list of different and eligible ops for this)
* [ ] Move Unary Ufuncs and general OpInfo tests (only ref checks) to a common place.
* [ ] Handle possible edge cases (for example: `uint64` isn't implemented in PyTorch but is there in NumPy, and this needs to be handled -- more on this later)
* [ ] Create a sample PR for a single (of all different categories?) on adding reference functions to the eligible ops.
* [ ] Add comments, remove unnecessary prints/comments (added for debugging).

Note: To keep the PR description short, examples of edge cases encountered have been mentioned in the comments below.

`cc: @mruberry @pmeier @kshitij12345`

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -146,6 +147,26 @@ def formatter(arg):
 
         return self._repr_helper(formatter)
 
+    def splat(self):
+        # TODO: Is this required? op(sample.splat()) doesn't work
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:151]:
```
======================================================================
ERROR: test_reference_testing_sub_test_subtract_cpu_int64 (__main__.TestOpInfoCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
    return test(*args, **kwargs)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 627, in wrapper
    fn(*args, **kwargs)
  File "/home/krshrimali/git/krshrimali/pytorch/test/test_ops.py", line 149, in test_reference_testing
    compare_with_reference(op, op.ref, sample_input)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 942, in compare_with_reference
    torch_fn(*sample_input.splat())
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_methods_invocations.py", line 331, in __call__
    return self.op(*args, **kwargs)
TypeError: sub() takes 2 positional arguments but 3 were given

----------------------------------------------------------------------
```

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -927,6 +927,34 @@ def get_comparison_dtype(a, b):
...
+def compare_with_reference(torch_fn, ref_fn, sample_input):
+    # Compares torch function with reference function for given sample input (object of SampleInput)
+    # Note: only values are compares, type comparison is not done here
+    try:
+        n_inp, n_args, _ = sample_input.numpy()
+    except TypeError:
+        return
+    t_inp, t_args, _ = sample_input.splat()
+
+    t_kwargs, n_kwargs = torch_fn.sample_kwargs(t_inp.device, None, t_inp)
```
Comment (krshrimali) [inline torch/testing/_internal/common_utils.py:940]: Example: `np.add()` doesn't take `alpha` as a keyword parameter. Is this the reason why this is required?

File: torch/testing/_internal/common_utils.py [hunk 3]
```diff
@@ -927,6 +927,34 @@ def get_comparison_dtype(a, b):
...
     return compare_dtype
 
+def compare_with_reference(torch_fn, ref_fn, sample_input):
+    # Compares torch function with reference function for given sample input (object of SampleInput)
+    # Note: only values are compares, type comparison is not done here
+    try:
+        n_inp, n_args, _ = sample_input.numpy()
+    except TypeError:
+        return
+    t_inp, t_args, _ = sample_input.splat()
```
Comment (krshrimali) [inline torch/testing/_internal/common_utils.py:938]: The output from `splat()` needs to be explicitly splatted like `torch_fn(t_input, *t_args, **t_kwargs)` else following error is raised if used as: `torch_fn(`sample_input.splat())`:

```
======================================================================
ERROR: test_reference_testing_sub_test_subtract_cpu_int64 (__main__.TestOpInfoCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
    return test(*args, **kwargs)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 627, in wrapper
    fn(*args, **kwargs)
  File "/home/krshrimali/git/krshrimali/pytorch/test/test_ops.py", line 149, in test_reference_testing
    compare_with_reference(op, op.ref, sample_input)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 942, in compare_with_reference
    torch_fn(*sample_input.splat())
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_methods_invocations.py", line 331, in __call__
    return self.op(*args, **kwargs)
TypeError: sub() takes 2 positional arguments but 3 were given

----------------------------------------------------------------------
```

**Review by krshrimali — COMMENTED** [review]
Tested for: `add, subtract, amax, argmax, roll` (numpy references)
To test: `linalg.svd` (scipy reference)

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -37,6 +37,7 @@
 from setuptools import distutils
 
 if TEST_SCIPY:
+    import scipy
     import scipy.special
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:41]: Okay, it's required. xD

**Comment by krshrimali** [comment]
```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 617, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 617, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 617, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 765, in only_fn
    return fn(self, device, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 640, in wrapper
    fn(*args, **kwargs)
  File "test_ops.py", line 58, in test_reference_testing
    compare_with_reference(op, op.ref, sample_input)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 970, in compare_with_reference
    torch.testing.assert_close(actual, torch.as_tensor(expected, dtype=actual.dtype), msg = msg)
AttributeError: 'torch.return_types.svd' object has no attribute 'dtype'
```
